File: Java/08.java
import java.util.HashMap;
import java.util.Locale;

public class Main {
    public static void main(String[] args) {
        // ORACIÓN DE PRUEBA: Incluye mayúsculas, minúsculas, puntuación y repeticiones.

        String oracionPrueba = "La casa azul, la casa verde. ¿Es esta la casa?";

        System.out.println("Oración de prueba: \"" + oracionPrueba + "\"");

        // LLAMADA A LA FUNCIÓN
        recuentoPalabras(oracionPrueba);

        System.out.println("\n--- OTROS EJEMPLOS ---");

        // Ejemplo 2: Más puntuación y más variaciones.
        String ejemplo2 = "Un Perro, dos Perros y TRES perros. ¡Perros!";
        System.out.println("Oración de prueba: \"" + ejemplo2 + "\"");
        recuentoPalabras(ejemplo2);
    }
    /**
     * Crea un programa que cuente cuantas veces se repite cada palabra
     * y que muestre el recuento final de todas ellas.
     * - Los signos de puntuación no forman parte de la palabra.
     * - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
     * - No se pueden utilizar funciones propias del lenguaje que
     *   lo resuelvan automáticamente.
     **/
    public static void recuentoPalabras(String oracion){
        oracion = oracion.replaceAll("[^\\p{Alnum}]", " ");
        oracion = oracion.toLowerCase(Locale.ROOT);
        String[] palabras = oracion.split("\\s+");
        HashMap<String, Integer> recuento = new HashMap<>();
        for (String palabra : palabras){
            if (recuento.containsKey(palabra)){
                recuento.put(palabra, recuento.get(palabra) + 1);
            } else {
                recuento.put(palabra, 1);
            }
        }
        for (String key : recuento.keySet()){
            System.out.println("La palabra "+ key + " aparece "+ recuento.get(key) + " veces");
        }
    }
}

File: C#/08.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace _08
{
    /*
     * Crea un programa que cuente cuantas veces se repite cada palabra
     * y que muestre el recuento final de todas ellas.
     * - Los signos de puntuación no forman parte de la palabra.
     * - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
     * - No se pueden utilizar funciones propias del lenguaje que
     *   lo resuelvan automáticamente.
     */
    public class ContadorPalabras
    {
        public static Dictionary<string, int> ContarPalabras(string oracion)
        {
            Dictionary<string, int> palabras = new Dictionary<string, int>();
            oracion = oracion.ToLower();
            oracion = Regex.Replace(oracion, @"[^\w\s@]+", "");
            string[] p = oracion.Split(' ');
            for (int i = 0; i < p.Length; i++)
            {
                if (!palabras.ContainsKey(p[i]))
                {
                    palabras.Add(p[i], 1);
                }
                else
                {
                    palabras[p[i]] = palabras[p[i]] + 1;
                }
            }
            return palabras;
        }
    }
}
